juju/model.py

```diff
@@ -2144,6 +2146,89 @@ async def _get_source_api(self, url, controller_name=None):
         await controller.connect(controller_name=controller_name)
         return controller
 
+    async def wait_for_idle(self, apps=None, raise_on_error=True, timeout=10 * 60,
+                            idle_period=15, check_freq=0.5):
+        """Wait for applications in the model to settle into an idle state.
+
+        :param apps (list[str]): Optional list of specific app names to wait on.
+            If given, all apps must be present in the model and idle, while other
+            apps in the model can still be busy. If not given, all apps currently
+            in the model must be idle.
+
+        :param raise_on_error (bool): If True, then any unit or app going into
+            "error" status immediately raises either a JujuAppError or a JujuUnitError.
+            Note that machine or agent failures will always raise an exception (either
+            JujuMachineError or JujuAgentError), regardless of this param.
+
+        :param timeout (float): How long to wait, in seconds, for the bundle settles
+            before raising an asyncio.TimeoutError. If None, will wait forever.
+
+        :param idle_period (float): How long, in seconds, the agent statuses of all
+            units of all apps need to be `idle`. This delay is used to ensure that
+            any pending hooks have a chance to start to avoid false positives.
+
+        :param check_freq (float): How frequently, in seconds, to check the model.
+        """
+        timeout = timedelta(seconds=timeout) if timeout is not None else None
+        idle_period = timedelta(seconds=idle_period)
+        start_time = datetime.now()
+        apps = apps or self.applications
+        idle_times = {}
+
+        def _raise_for_errors(errors):
+            if not errors:
+                return
+            for entity_name, error_type in (("Machine", JujuMachineError),
+                                            ("Agent", JujuAgentError),
+                                            ("Unit", JujuUnitError),
+                                            ("App", JujuAppError)):
+                errored = errors.get(entity_name, [])
+                if not errored:
+                    continue
+                raise error_type("{}{} in error: {}".format(
+                    entity_name,
+                    "s" if len(errored) > 1 else "",
+                    ", ".join(errored),
+                ))
+
+        while True:
+            busy_apps = []
+            errors = {}
+            for app in apps:
+                if app not in self.applications:
+                    busy_apps.append(app)
+                    continue
+                app = self.applications[app]
+                if raise_on_error and app.status == "error":
```

Comment:
On most new controllers this will probably just be `unset` unless the charm explicitly sets the status, so I'm glad we also check the units as well :+1: